platform/ios/src/MGLAnnotationView.mm

```diff
@@ -212,13 +212,21 @@ - (void)setDragState:(MGLAnnotationViewDragState)dragState animated:(BOOL)animat
         [self.mapView.calloutViewForSelectedAnnotation dismissCalloutAnimated:animated];
         [self.superview bringSubviewToFront:self];
     }
-
-    if (dragState == MGLAnnotationViewDragStateEnding)
+    else if (dragState == MGLAnnotationViewDragStateCanceling)
+    {
+        self.panGestureRecognizer.enabled = NO;
+        self.longPressRecognizer.enabled = NO;
+        self.center = [self.mapView convertCoordinate:self.annotation.coordinate toPointToView:self.mapView];
+        self.panGestureRecognizer.enabled = YES;
+        self.longPressRecognizer.enabled = YES;
+        self.dragState = MGLAnnotationViewDragStateNone;
+    }
+    else if (dragState == MGLAnnotationViewDragStateEnding)
     {
-        if ([self.mapView.delegate respondsToSelector:@selector(mapView:didDragAnnotationView:toCoordinate:)])
+        if ([self.annotation respondsToSelector:@selector(setCoordinate:)])
         {
             CLLocationCoordinate2D coordinate = [self.mapView convertPoint:self.center toCoordinateFromView:self.mapView];
-            [self.mapView.delegate mapView:self.mapView didDragAnnotationView:self toCoordinate:coordinate];
+            [(NSObject *)self.annotation setValue:[NSValue valueWithMGLCoordinate:coordinate] forKey:@"coordinate"];
```

Comment:
Right, fortunately, we can assume Objective-C dispatch even in Swift, thanks to the MGLAnnotation protocol. I think this KVO approach will be fine as long as the class doesn’t do any fancy KVO tricks to forward the `coordinate` key path to a selector other than `-setCoordinate:`.

Theoretically, if it’s an NSProxy that happens to respond to `-setValue:forKey:`, casting to `NSObject *` could be problematic, but that’s really going off into asking-for-trouble land.